### ISSUE TITLE
first PDF call does not work (NullPointerException or missing images)

### ISSUE THREAD

**Issue by voisardf** [openoereb/pyramid_oereb#239]
After starting the sample environnement with make serve, then calling the PDF extract for the first time (full or reduced) with 
http://localhost:6543/oereb/extract/full/pdf/CH113928077734 (in my case)

I get a 
`Error while processing request:
java.lang.NullPointerException`
first. On the second try I get the PDF as expected

**Comment by sbrunner** [comment]
I suspect that's an issue in mapfish print...

**Comment by sbrunner** [comment]
Exception:
```
print_1  | 08:56:38.014 [ForkJoinPool-2-worker-3] DEBUG o.m.p.processor.ProcessorGraphNode - Succeeded in executing process: SetFeaturesProcessor
print_1  | 08:56:38.014 [ForkJoinPool-2-worker-23] DEBUG o.m.p.processor.ProcessorGraphNode - Succeeded in executing process: SetFeaturesProcessor
print_1  | 08:56:38.014 [ForkJoinPool-2-worker-23] INFO  o.m.p.processor.ProcessorGraphNode - Time taken to run processor: 'class org.mapfish.print.processor.map.SetFeaturesProcessor' was 130 ms
print_1  | 08:56:38.015 [ForkJoinPool-2-worker-3] INFO  o.m.p.processor.ProcessorGraphNode - Time taken to run processor: 'class org.mapfish.print.processor.map.SetFeaturesProcessor' was 130 ms
print_1  | 08:56:38.015 [ForkJoinPool-2-worker-23] DEBUG o.m.p.processor.ProcessorGraphNode - Executing process: CreateMapProcessor
print_1  | 08:56:38.016 [ForkJoinPool-2-worker-3] DEBUG o.m.p.processor.ProcessorGraphNode - Executing process: CreateMapProcessor
print_1  | 08:56:38.056 [ForkJoinPool-2-worker-23] ERROR o.m.p.processor.ProcessorGraphNode - Error while executing process: CreateMapProcessor
print_1  | java.lang.NullPointerException: null
print_1  |      at org.mapfish.print.processor.map.CreateMapProcessor.zoomToFeatures(CreateMapProcessor.java:581)
print_1  |      at org.mapfish.print.processor.map.CreateMapProcessor.execute(CreateMapProcessor.java:171)
print_1  |      at org.mapfish.print.processor.map.CreateMapProcessor.execute(CreateMapProcessor.java:104)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:200)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.executeDependencyProcessors(ProcessorGraphNode.java:251)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:226)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.executeDependencyProcessors(ProcessorGraphNode.java:251)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:226)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.executeDependencyProcessors(ProcessorGraphNode.java:251)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:226)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.executeDependencyProcessors(ProcessorGraphNode.java:251)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:226)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:176)
print_1  |      at jsr166y.RecursiveTask.exec(RecursiveTask.java:64)
print_1  |      at jsr166y.ForkJoinTask.doExec(ForkJoinTask.java:305)
print_1  |      at jsr166y.ForkJoinWorkerThread.execTask(ForkJoinWorkerThread.java:575)
print_1  |      at jsr166y.ForkJoinPool.scan(ForkJoinPool.java:733)
print_1  |      at jsr166y.ForkJoinPool.work(ForkJoinPool.java:617)
print_1  |      at jsr166y.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:369)
print_1  | 08:56:38.059 [ForkJoinPool-2-worker-23] INFO  o.m.p.processor.ProcessorGraphNode - Time taken to run processor: 'class org.mapfish.print.processor.map.CreateMapProcessor' was 43 ms
print_1  | 08:56:38.075 [ForkJoinPool-2-worker-13] DEBUG o.m.p.p.ProcessorDependencyGraph - Finished executing processor graph: 
print_1  | + DataSourceProcessor
print_1  | + DataSourceProcessor
print_1  | + DataSourceProcessor
print_1  | + DataSourceProcessor
print_1  | + AddStaticLayersProcessor
print_1  |   +-- AddStaticLayersProcessor
print_1  |     +-- SetStyleProcessor
print_1  |       +-- SetFeaturesProcessor
print_1  |         +-- CreateMapProcessor
print_1  | 08:56:38.076 [ForkJoinPool-2-worker-11] ERROR o.m.p.processor.ProcessorGraphNode - Error while executing process: DataSourceProcessor
print_1  | java.lang.NullPointerException: null
print_1  |      at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
print_1  |      at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
print_1  |      at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
print_1  |      at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
print_1  |      at jsr166y.ForkJoinTask.getThrowableException(ForkJoinTask.java:507)
print_1  |      at jsr166y.ForkJoinTask.reportResult(ForkJoinTask.java:567)
print_1  |      at jsr166y.ForkJoinTask.join(ForkJoinTask.java:611)
print_1  |      at org.mapfish.print.processor.jasper.DataSourceProcessor.processInput(DataSourceProcessor.java:243)
print_1  |      at org.mapfish.print.processor.jasper.DataSourceProcessor.execute(DataSourceProcessor.java:201)
print_1  |      at org.mapfish.print.processor.jasper.DataSourceProcessor.execute(DataSourceProcessor.java:75)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:200)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:176)
print_1  |      at jsr166y.RecursiveTask.exec(RecursiveTask.java:64)
print_1  |      at jsr166y.ForkJoinTask.doExec(ForkJoinTask.java:305)
print_1  |      at jsr166y.ForkJoinWorkerThread.execTask(ForkJoinWorkerThread.java:575)
print_1  |      at jsr166y.ForkJoinPool.scan(ForkJoinPool.java:733)
print_1  |      at jsr166y.ForkJoinPool.work(ForkJoinPool.java:617)
print_1  |      at jsr166y.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:369)
print_1  | Caused by: java.lang.NullPointerException: null
print_1  |      at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
print_1  |      at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
print_1  |      at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
print_1  |      at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
print_1  |      at jsr166y.ForkJoinTask.getThrowableException(ForkJoinTask.java:507)
print_1  |      at jsr166y.ForkJoinTask.reportResult(ForkJoinTask.java:567)
print_1  |      at jsr166y.ForkJoinTask.join(ForkJoinTask.java:611)
print_1  |      at org.mapfish.print.processor.ProcessorDependencyGraph$ProcessorGraphForkJoinTask.compute(ProcessorDependencyGraph.java:185)
print_1  |      at org.mapfish.print.processor.ProcessorDependencyGraph$ProcessorGraphForkJoinTask.compute(ProcessorDependencyGraph.java:140)
print_1  |      ... 6 common frames omitted
print_1  | Caused by: java.lang.NullPointerException: null
print_1  |      at org.mapfish.print.processor.map.CreateMapProcessor.zoomToFeatures(CreateMapProcessor.java:581)
print_1  |      at org.mapfish.print.processor.map.CreateMapProcessor.execute(CreateMapProcessor.java:171)
print_1  |      at org.mapfish.print.processor.map.CreateMapProcessor.execute(CreateMapProcessor.java:104)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:200)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.executeDependencyProcessors(ProcessorGraphNode.java:251)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:226)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.executeDependencyProcessors(ProcessorGraphNode.java:251)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:226)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.executeDependencyProcessors(ProcessorGraphNode.java:251)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:226)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.executeDependencyProcessors(ProcessorGraphNode.java:251)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:226)
print_1  |      at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:176)
print_1  |      ... 6 common frames omitted
print_1  | 08:56:38.077 [ForkJoinPool-2-worker-11] INFO  o.m.p.processor.ProcessorGraphNode - Time taken to run processor: 'class org.mapfish.print.processor.jasper.DataSourceProcessor' was 4955 ms
```

**Comment by ger-benjamin** [comment]
It is still open ? I just realize that I can't reproduce this issue anymore. I don't know why. I'm near on the latest commit (d20f1c0e13f5f536171f5bcf0d89224b38ff2d52).

Previously, I was surprised to see this issue disappear when I remove the link with the Report.properties (https://github.com/camptocamp/pyramid_oereb/blob/master/print/print-apps/oereb/config.yaml#L3) But perhaps that was a coincidence.

**Comment by voisardf** [comment]
I still got this problem sometimes, but @kalbermattenm thinks that this could be due to the long initialization process of ApacheTomcat 8.5 - I keep an eye on that to confirm or not...

**Comment by kalbermattenm** [comment]
One simple test you could do is to completely empty your webapps folder only leaving the war file related to pyramid_orereb project, then start Tomcat and check if it gets better....

**Comment by voisardf** [comment]
As of today I still get the NullPointerException some times and it's not related to the startup time of Tomcat but to MapfishPrint as it seems - here a extract of the log:

```
gradle: 09:47:33.657 [ForkJoinPool-2-worker-36] ERROR o.m.p.processor.ProcessorGraphNode - Error while executing process: CreateMapProcessor
java.lang.NullPointerException: null
	at org.mapfish.print.processor.map.CreateMapProcessor.zoomToFeatures(CreateMapProcessor.java:581)
	at org.mapfish.print.processor.map.CreateMapProcessor.execute(CreateMapProcessor.java:171)
	at org.mapfish.print.processor.map.CreateMapProcessor.execute(CreateMapProcessor.java:104)
	at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:200)
	at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.executeDependencyProcessors(ProcessorGraphNode.java:251)
	at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:226)
	at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.executeDependencyProcessors(ProcessorGraphNode.java:251)
	at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:226)
	at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.executeDependencyProcessors(ProcessorGraphNode.java:251)
	at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:226)
	at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.executeDependencyProcessors(ProcessorGraphNode.java:251)
	at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:226)
	at org.mapfish.print.processor.ProcessorGraphNode$ProcessorNodeForkJoinTask.compute(ProcessorGraphNode.java:176)
	at jsr166y.RecursiveTask.exec(RecursiveTask.java:64)
	at jsr166y.ForkJoinTask.doExec(ForkJoinTask.java:305)
	at jsr166y.ForkJoinWorkerThread.execTask(ForkJoinWorkerThread.java:575)
	at jsr166y.ForkJoinPool.scan(ForkJoinPool.java:733)
	at jsr166y.ForkJoinPool.work(ForkJoinPool.java:617)
	at jsr166y.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:369)
```

**Comment by ralphschulerSZ** [comment]
I can confirm the NullPointerException can sometimes appear during the normal running of Tomcat and Apache.

Here is an extract of the Apache log:
```

[Wed Jan 10 07:55:42.216819 2018] [wsgi:error] [pid 26712] 2018-01-10 07:55:42,216 WARNI [py.warnings][MainThread] /srv/apache/schwyz-geoportal/private/geoportal/.build/venv/lib/python2.7/site-packages/sqlalchemy/engine/default.py:585: SAWarning: Unicode type received non-unicodebind param value.
[Wed Jan 10 07:55:42.216846 2018] [wsgi:error] [pid 26712]   processors[key](compiled_params[key])
[Wed Jan 10 07:55:42.216849 2018] [wsgi:error] [pid 26712]
[Wed Jan 10 07:56:19.971527 2018] [wsgi:error] [pid 26714] /srv/apache/schwyz-geoportal/private/schwyz_oereb/.venv/lib/python2.7/site-packages/pyramid_oereb/lib/records/documents.py:126: UserWarning: Type of "abbreviation" should be "dict"
[Wed Jan 10 07:56:19.971552 2018] [wsgi:error] [pid 26714]   warnings.warn('Type of "abbreviation" should be "dict"')
```

**Comment by lopo977** [comment]
Same problem here, I keep getting the null pointer exception by the first request.
The problem is related to the insufficient time to load/compile the jasper report isn't it possible to give more time to the process or trigger the completion of it before trying to read it? Ideally, I would compile it by the bootstrap of the tomcat server.

**Comment by jwkaltz** [comment]
Update on this issue:
While analyzing this behaviour, a bug was found in mapfish-print and addressed in https://github.com/mapfish/mapfish-print/pull/656. Unfortunately, this does not fully solve the problem. When using mfp 3.12.1, on the first call for the pyramid_oereb extract pdf, you no longer get a NullPointerException, but some of the images are missing (which images are missing varies between restarts). On subsequent calls, all images are included as expected.

**Comment by vvmruder** [comment]
Regarding to response from @voisardf this seems to happen only while tomcat is starting. Once it is fully up and running it seems to work without a problem.

First tests in our environment seems to proof this.

**Comment by jwkaltz** [comment]
Issue opened in mapfish-print open source project: https://github.com/mapfish/mapfish-print/issues/819

**Comment by pvalsecc** [comment]
I've released MapFish print 3.10.8 and 3.18.2 with the fix. Can somebody please test?
On my laptop, I cannot reproduce anymore.

**Comment by jwkaltz** [comment]
Thanks Patrick - I will test together with @ralphschulerSZ, but not before tuesday.
@voisardf @lopo977 maybe you can verify on your server installations?

**Comment by voisardf** [comment]
@jwkaltz : Any particular release or tag to use for testing?

**Comment by jwkaltz** [comment]
@voisardf yes, 3.10.8 if you are still using a 3.10 version, or 3.18.2 if you are already using a newer version.

**Comment by jwkaltz** [comment]
@voisardf I was now able to test on a real server environment. With the update to 3.18.2, the problem is gone. As the same fix was backported to 3.10.8, I am confident it will be solved for 3.10 users as well (as long as they update to 3.10.8.)

**Comment by lopo977** [comment]
The problem is gone also in our installations. Thank you!
